protocol/l2-forked-tests.md

```diff
@@ -0,0 +1,209 @@
+# Forked Testing Framework for L2 Networks: Design Doc
+
+## Purpose
+
+This document outlines the proposed design for an integrated L2 Forked Test suite and defines the scope of this integration.
+
+## Summary
+
+This proposal describes a testing framework for L2 networks that lets us run tests against realistic contract states. The framework applies Network Upgrade Transactions (NUTs) to forked networks, creating a consistent environment for testing Predeploy contracts. By implementing NUTs in Solidity or combining Go with Solidity, we can improve test coverage, build confidence in our upgrades, and identify issues earlier.
+
+## Problem Statement + Context
+
+As Predeploy contracts development advances, we currently lack an efficient method to run test suites against the state of arbitrary networks. We need to develop a configurable testing framework that accurately replicates the state of Predeploy contracts within selected networks, initialized from a specified block number (initial state).
+
+This framework should be able to take a network's initial state and apply the required Network Upgrade Transactions (NUTs) to reach the latest version of the contracts for running the test suite. This approach would help increase confidence that the upgrade process doesn't introduce unexpected bugs, allowing contributors to catch errors earlier in the release process. While this implementation won't perfectly mirror the environment after NUTs are applied to the real network, it represents a significant improvement to our testing capabilities.
+
+The solution must provide a deterministic state to the test suite, enabling the execution of targeted test cases against this final state. This state results from applying a set of NUTs to the initial forked state. Therefore, we need to implement a structured way to describe these NUTs and a method to apply them.
+
+It's also important that this solution remains loosely coupled with the test setup itself, so that running tests against different chains starting with different initial states requires minimal changes to the Setup test file.
+
+## Proposed Solution
+
+Currently, L1 forked state is abstracted by ForkLive, which handles reading the L1 contract's code and upgrading when needed. Ideally, we'd like a similar approach for L2 fork tests, where a single contract is responsible for setting up L2 Predeploys in the final state needed for our test suite. We propose a two-part approach: one part abstracts the NUTs that occur in each upgrade, while the other part executes them sequentially. For L2, we have the advantage of knowing the addresses in advance, so unlike ForkLive, storing addresses isn't necessary.
+
+### Implementation Components
+
+1. **NUT Definitions**: Specifications of the Network Upgrade Tests to be executed prior to test suite initialization
+
+```solidity
+interface NUTExecutor {
+    function execute(bytes calldata _calldata) external;
```

Comment:
how is `_calldata` defined? 
Do we expect that it might need to vary for different networks, or do we maybe not need it?